### ISSUE TITLE
Processing of a lookup with multiple subtables isn't described

### ISSUE THREAD

**Issue by PeterConstable** [MicrosoftDocs/typography-issues#1142]
The spec currently describes how a lookup is processed:

>A lookup specifies an input sequence pattern of one or more glyphs. When the lookup is processed over a glyph sequence within a string, the client starts with the first glyph in the sequence, testing for a match with the input sequence pattern specified by the lookup. If the glyph sequence does not match the lookup input sequence pattern, processing of the lookup is finished for that glyph, and the client advances to the next glyph in the glyph sequence. If the glyph sequence does match the lookup input sequence pattern, then the substitution or position operation is performed on the matched input sequence. A lookup is finished for a glyph after the client has performed the substitution or positioning operation.

This doesn't make clear what is done if the lookup has multiple lookup subtables. _Are all subtables processed using the same starting glyph position?_ (Not the case.) _Are they processed in order until the input sequence pattern of a subtable is matched, then processing of the lookup stops after that subtable?_ (Yes!)

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: 1cb42967-20ca-9270-4572-65315127fcc5
* Version Independent ID: 1fa413f1-71cf-ac2a-0999-70dd5d8ac4cd
* Content: [OpenType layout common table formats (OpenType 1.9) - Typography](https://learn.microsoft.com/en-us/typography/opentype/spec/chapter2#classDefTbl)
* Content Source: [typographydocs/opentype/spec/chapter2.md](https://github.com/MicrosoftDocs/typography/blob/live/typographydocs/opentype/spec/chapter2.md)
* Service: **typography**
* GitHub Login: @PeterCon
* Microsoft Alias: **alib**

**Comment by PeterCon** [comment]
Added a paragraph explaiing this:

>If a Lookup table has multiple subtables, the subtables are processed in order, testing the glyph sequence at the current glyph position for a match with the input sequence patterns specified by each subtable in turn. If there is no match with the pattern of a subtable, then processing moves to the next subtable. If the glyph sequence does not match the patterns of any lookup subtable, processing of the lookup is finished for that glyph position. If the glyph sequence does match the pattern of a subtable, then the operations of that subtable are performed and processing of the lookup is finished—no further subtables are processed for that position in the glyph sequence.

See the [OT 1.9.1 alpha](https://docs.microsoft.com/en-us/typography/opentype/otspec191alpha/ot191alpha) for draft revisions addressing this issue.

**Comment by PeterCon** [comment]
I contemplated whether to add a note explaining the possibility of doing what VOLT does with "EXCEPT" contexts: it may be easier for the font developer to describe contexts in which an action should _not_ be done, and that can be implemented in a lookup by having the initial subtable(s) describe the specific _exception_ context(s) but perform no action, and then have following subtable(s) perform actions on unrestricted contexts. But that feels like it might be going beyond the scope of the spec.

**Comment by tiroj** [comment]
Yes, I think that is going beyond the spec. Clearly stating how subtables are processed is sufficient, I think, and then font makers can decide how best to use this information, which may vary across fonts. There are times when I use subtables to process exceptions and skip more general substitutions or positioning, and times when I prefer to apply exceptions as separate lookups after the general subsitutions or positioning.

[On the subject of clarifications of how OTL is processed, I can’t recall if there is anything in the spec that points out that GPOS adjustments are additive (adjustments in subsequent lookups are added to those made in previous lookups), while GPOS anchoring is substitutive (anchors in subsequent lookups override those in previous lookups). This is something I have to explain to font makers, and wonder if it is clearly stated somewhere to which I could point?]

**Comment by PeterCon** [comment]
For OT 1.9.1, I've revised a paragraph in the [OTL Common Table Formats](https://learn.microsoft.com/en-us/typography/opentype/otspec191alpha/chapter2) chapter that relates to that:

>During text processing, a client applies a feature to some sequence of glyphs for a string. It then processes the lookups referenced by that feature in their lookup list order. For each lookup, the client processes that lookup over each glyph in the sequence to which the feature has been applied. After that lookup has been processed <del>for each glyph in the</del><ins>over the entire glyph</ins> sequence, it then processes the next lookup referenced by the feature in the same manner. <ins>As each lookup is processed, it acts on the substitution or positioning results of the previous lookups.</ins> This continues until all lookups referenced by the feature have been processed.

There were similar statements in some other places but it makes sense to say that in the section on lookups.

As for anchors being substitutive, that feels like it would belong in the GPOS chapter, in sections that describe attachment positioning. You might want to open an issue on the GPOS chapter regarding that.

**Comment by PeterCon** [comment]
@tiroj In the [draft GPOS chapter for OT 1.9.1 alpha](https://learn.microsoft.com/en-us/typography/opentype/otspec191alpha/gpos), I revised an existing paragraph:

>A series of positioning operations on the same glyph or string requires multiple lookups, one for each separate action. <del>Positioning adjustment values are accumulated in these cases. </del>Each lookup has a different array index in the LookupList table and is applied in the LookupList order. <ins>The positioning adjustment of each lookup is applied to the result of previous lookups. When the adjustments are expressed as absolute placement or advance adjustments in the X or Y direction, these adjustments are accumulated as each lookup is processed. For adjustments expressed using attachment points, however, attachment point positioning can override the effect of preceding lookups.</ins>

**Comment by tiroj** [comment]
Looks good. I can’t think of any simpler way to express it.

**Comment by PeterCon** [comment]
This is fixed in OT 1.9.1. See the [OT 1.9.1 beta](https://docs.microsoft.com/en-us/typography/opentype/otspec191/ot191beta) for draft revisions addressing this issue. Closing.
